src/composer.ts

```diff
@@ -34,6 +34,14 @@ function getText(
   return undefined
 }
 
+type MatchedContext<
```

Comment:
Does this help you in any way?

If you were thinking of having it very differently, feel free to change it as you like

Reply:
Nope. I need a type with params `<C extends Context, U extends tt.Update>`, independent of `tt.UpdateType | tt.MessageSubType`.

> feel free to change it as you like

I can't. I don't understand your types. They're complex, and there's so many of them. I'd need to ignore or scrap them and start from scratch. Hence why I'm asking you to give me some entry point I can immediately use, and which would perhaps allow me to understand and simplify rest of your types.

Reply:
> I can't. I don't understand your types.

I see. Did you see https://github.com/telegraf/telegraf/pull/1195#discussion_r532059517? Don't hesitate to ask, I am aware that what I did is not straightforward at all

Reply:
I did. But there's so much stuff that I don't know where to even begin to look!

In my opinion, only 3 types are needed: `MountMap`, which maps `tt.UpdateType | tt.MessageSubType` into a `TUpdate`, `GuardedContexts` which intersects `TContext` with a type derieved from `TUpdate`, and one helper.

Reply:
> I don't know where to even begin to look

`MatchedContext` is where everything begins. I wrote some docs now.

It takes a context and makes those things required that should be required. It makes those things undefined that should be undefined.

It does this by doing an intersection with some objects that have very specific properties set.

Reply:
Here is an illustration of this. I hope it addresses the confusion.

```
type Context = {
  message: {
    text: string | undefined
    left_chat_member: true | undefined
    other_prop: symbol // or whatever
  }
}

type GuaranteedStuff = {
    message: {
      text: string
      // we ignore all the rest here …
  }
}
type UndefinedStuff = {
    message: {
      left_chat_member: undefined
    }
}

// what we actually want to have in the end!
type MatchedContext = Context & GuaranteedStuff & UndefinedStuff
```

Reply:
Thanks, this helps a lot! I wish types you added to `telegram-types.ts` received the same treatment.

Reply:
Your wish is my command.

Reply:
Looking at your suggestion for `guard`, I think I see what you were expecting from the types I wrote, and why this did not match. You needed some sort of type guarding logic for different update types (`update is TUpdate`), while I implemented a type transformation that narrows down some unions deep down in a context object. Those approached look fundamentally different to me. Maybe it's possible to somehow make one use the other, but I don't really know how.